### PR TITLE
Finished tilemap generation and stl and made it jit-able

### PR DESCRIPTION
**Pull request by benbenwsh**
* Translated C# WaveFunctionCollapse library to Python
* Made all functions and classes jit-able
* Take in the generated tilemap and convert it to stl format (yet to optimise)
* Wrote tests for XML Reader and JIT Model (tilemap generation)

### CODE REVIEW

File: src/icland/world_gen/JITModel.py [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+
+    return cast(
+        tuple[JITModel, jax.Array],
+        jax.lax.while_loop(condition, body_func, (model, success))[0],
+    )
+
+
+model = sample_world(10, 10, 1000, jax.random.key(42), True, 1)
+one_hot = export(model, TILECODES, 10, 10)
+print(one_hot.tolist())
```
Comment (harens) [inline src/icland/world_gen/JITModel.py]: This should be wrapped in a `if __name__ == "__main__"` so that it's not run when we import this file.

File: src/icland/world_gen/JITModel.py [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+    propagator: jax.Array  # (4, T, propagator_length)
+    distribution: jax.Array  # shape: (T, )
+
+    key: jax.Array
+
+
+@partial(jax.jit, static_argnums=[0, 1, 2])
+def _init(
+    width: jnp.int32,
+    height: jnp.int32,
```
Comment (harens) [inline src/icland/world_gen/JITModel.py:111]: What was the reason why width and height are static args (partial jax)? I didn't notice any conditional statements with these vars, but I might have missed something.

File: src/icland/world_gen/JITModel.py [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+                    model,
+                )
+                return model
+
+            model = jax.lax.fori_loop(
+                0, model.wave.shape[0], handle_completion_inner, model
+            )
+            return model, jnp.array(True, dtype=bool), jnp.array(True, dtype=bool)
+
+        print("Node: ", node)
```
Comment (harens) [inline src/icland/world_gen/JITModel.py]: If we need to keep this print, we can use python's debug library? Otherwise, let's remove it.

Same is true for the other prints.

File: src/icland/world_gen/JITModel.py [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+
+    # Define the loop state
+    model = _clear(model)
+    init_state = {"model": model, "steps": 0, "done": False, "success": True}
+
+    def cond_fun(state: dict[str, Any]) -> Any:
+        """Condition function for the while loop."""
+        return (~state["done"]) & (state["steps"] < max_steps)
+
+    def body_fun(state: dict[str, Any]) -> dict[str, Any]:
```
Comment (harens) [inline src/icland/world_gen/JITModel.py]: Ideally we try to reduce the use of [Any](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#the-any-type) type.

We could use [TypedDict](https://mypy.readthedocs.io/en/stable/typed_dict.html) here rather than `Any` since we know the exact types of the arguments (JITModel, jax.Array, jax.Array).

File: src/icland/world_gen/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+numpy
```
Comment (harens) [inline src/icland/world_gen/requirements.txt]: Would it be possible to remove the requirements.txt in favour of a unified pyproject.toml file?

File: src/icland/world_gen/tile_data.py [hunk 6]
```diff
@@ -0,0 +1,1177 @@
+"""Static data used for Wave Function Collapse algorithm."""
```
Comment (harens) [inline src/icland/world_gen/tile_data.py:1]: This file has over 1000 lines of code! I'm wondering if it's better to store the data in a serialised format to improve memory usage and performance.

File: tests/test_jit_model.py [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+            1,
+        ),  # rand_value=1 should select index 1 (should never actually return 1, but we're testing edge cases)
+    ],
+)
+def test_random_index_from_distribution(
+    distribution: jax.Array, rand_value: jax.Array, expected_result: jax.Array
+) -> None:
+    """Test the random_index_from_distribution function."""
+    # JIT compile the function
+    jit_func = jax.jit(_random_index_from_distribution)
```
Comment (harens) [inline tests/test_jit_model.py:81]: Wouldn't this keep recompiling the jitted version of the function every time we run with different parameters? I would've thought that we compile outside the function.

**Review by harens — COMMENTED** [review]
Really great work, thank you! 🎉 Some minor comments from me.

File: src/icland/world_gen/JITModel.py (hunk 2 above)
Comment (Iajedi) [reply src/icland/world_gen/JITModel.py:111]: `width` and `height` are used to initialise a majority of the arrays e.g. `compatible` etc. in the `JITModel` object, and since array sizes must be "static" at compile time they are defined as static args, which is fine since across the worlds we wouldn't need to adjust it anyway at this stage (unless we would like to support this functionality).

File: src/icland/world_gen/requirements.txt (hunk 5 above)
Comment (Iajedi) [reply src/icland/world_gen/requirements.txt]: Paging @benbenwsh.

File: src/icland/world_gen/tile_data.py (hunk 6 above)
Comment (Iajedi) [reply src/icland/world_gen/tile_data.py:1]: Sure, this can be done by `np.save`. The problem here (that needs to be resolved when we integrate this into a JITted pipeline in the future) is how do we ingest/read using IO a file from disk in a JITted function? Please advise on this matter.

File: tests/test_jit_model.py (hunk 7 above)
Comment (Iajedi) [reply tests/test_jit_model.py:81]: paging @benbenwsh

File: src/icland/world_gen/JITModel.py (hunk 2 above)
Comment (harens) [reply src/icland/world_gen/JITModel.py:111]: Thank you for clarifying Isaac!

File: src/icland/world_gen/tile_data.py (hunk 6 above)
Comment (harens) [reply src/icland/world_gen/tile_data.py:1]: Ah good point. I was thinking `jax.numpy.save`, but then noticed [this](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.load.html):

> This function requires concrete array inputs, and is not compatible with transformations like [jax.jit()](https://jax.readthedocs.io/en/latest/_autosummary/jax.jit.html#jax.jit) or [jax.vmap()](https://jax.readthedocs.io/en/latest/_autosummary/jax.vmap.html#jax.vmap).

Only pure functions are jittable, so maybe this is ok then? Thanks Isaac.

File: pyproject.toml [hunk 8]
```diff
@@ -6,10 +6,20 @@ readme = "README.md"
...
     "imageio[ffmpeg]>=2.37.0",
-    "jax[cuda12]>=0.4.38",
+    "jax>=0.5.0 ; sys_platform != 'linux'",
+    "jax[cuda12]>=0.5.0 ; sys_platform == 'linux'",
+    "matplotlib>=3.10.0",
     "mujoco>=3.2.6",
     "mujoco-mjx>=3.2.6",
+    "numpy>=2.2.2",
+    "numpy-stl>=3.2.0",
+    "pytest>=8.3.4",
```
Comment (harens) [inline pyproject.toml]: Are all these dependencies needed for users to run the project? pytest is already a developer dependency, and we don't need sphinx (that's a documentation library). Not sure about the others.

Comment (Iajedi) [reply pyproject.toml]: Right. I removed `sphinx` and `pytest` (not sure why sphinx was here in the first place), but kept the others, since numpy-stl is still being used to export to Gursukh's end.

Comment (harens) [reply pyproject.toml]: We shouldn't need `types-setuptools`. I'm guessing matplotlib and scipy are being used somewhere I can't see?

**Review by harens — APPROVED** [review]
Thank you world gen team! LGTM once the unresolved comments are handled. 🎉
